libsel4/tools/syscall_stub_gen_rs.py

```diff
@@ -0,0 +1,776 @@
+            CapType("seL4_X86_Page", wordsize),
+            CapType("seL4_X86_PageDirectory", wordsize),
+            CapType("seL4_X86_PageTable", wordsize),
+            CapType("seL4_X86_IOPageTable", wordsize),
+            CapType("seL4_X86_VCPU", wordsize),
+            CapType("seL4_X86_EPTPML4", wordsize),
+            CapType("seL4_X86_EPTPDPT", wordsize),
+            CapType("seL4_X86_EPTPD", wordsize),
+            CapType("seL4_X86_EPTPT", wordsize),
+            StructType("seL4_VCPUContext", wordsize * 7, wordsize),
+            StructType("seL4_UserContext", wordsize * 12, wordsize),
+        ],
+
+        "x86_64": [
+            Type("seL4_X86_VMAttributes", wordsize, wordsize),
+            CapType("seL4_X86_IOPort", wordsize),
+            CapType("seL4_X86_IOPortControl", wordsize),
+            CapType("seL4_X86_ASIDControl", wordsize),
+            CapType("seL4_X86_ASIDPool", wordsize),
+            CapType("seL4_X86_IOSpace", wordsize),
+            CapType("seL4_X86_Page", wordsize),
+            CapType("seL4_X64_PML4", wordsize),
+            CapType("seL4_X86_PDPT", wordsize),
+            CapType("seL4_X86_PageDirectory", wordsize),
+            CapType("seL4_X86_PageTable", wordsize),
+            CapType("seL4_X86_IOPageTable", wordsize),
+            CapType("seL4_X86_VCPU", wordsize),
+            CapType("seL4_X86_EPTPML4", wordsize),
+            CapType("seL4_X86_EPTPDPT", wordsize),
+            CapType("seL4_X86_EPTPD", wordsize),
+            CapType("seL4_X86_EPTPT", wordsize),
+            StructType("seL4_VCPUContext", wordsize * 7, wordsize),
+            StructType("seL4_UserContext", wordsize * 20, wordsize),
+        ],
+        "riscv32": [
+            Type("seL4_RISCV_VMAttributes", wordsize, wordsize),
+            CapType("seL4_RISCV_Page", wordsize),
+            CapType("seL4_RISCV_PageTable", wordsize),
+            CapType("seL4_RISCV_ASIDControl", wordsize),
+            CapType("seL4_RISCV_ASIDPool", wordsize),
+            StructType("seL4_UserContext", wordsize * 32, wordsize),
+        ],
+        "riscv64": [
+            Type("seL4_RISCV_VMAttributes", wordsize, wordsize),
+            CapType("seL4_RISCV_Page", wordsize),
+            CapType("seL4_RISCV_PageTable", wordsize),
+            CapType("seL4_RISCV_ASIDControl", wordsize),
+            CapType("seL4_RISCV_ASIDPool", wordsize),
+            StructType("seL4_UserContext", wordsize * 32, wordsize),
+        ]
+    }
+    # legacy alias
+    arch_types["arm_hyp"] = arch_types["aarch32"]
+    return arch_types
+
+
+def generate_param_list(input_params, output_params):
+    # Generate parameters
+    params = []
+    for param in input_params:
+        t = param.type
+        if t.pass_by_reference():
+            t = t.asPointerType(False)
+        params.append(t.render_parameter_name(param.name))
+    for param in output_params:
+        t = param.type
+        if t.pass_by_reference():
```

Comment:
make this an assert here?
```
assert t.pass_by_reference()
```

Reply:
When we have `if param.type.pass_by_reference():` the code is generated correctly.
If assert it here, the script will fail even though there's no need for that (We just need to skip unneeded arguments).
Here are some generated snippets to demonstrate what happens:

C code:
```
LIBSEL4_INLINE seL4_TCB_GetBreakpoint_t
seL4_TCB_GetBreakpoint(seL4_TCB _service, seL4_Uint16 bp_num)
{
	seL4_TCB_GetBreakpoint_t result;
	seL4_MessageInfo_t tag = seL4_MessageInfo_new(TCBGetBreakpoint, 0, 0, 1);
...
```

Rust generated as expected:
```
#[inline(always)]
pub unsafe fn seL4_TCB_GetBreakpoint(_service: seL4_TCB, bp_num: u16) -> seL4_TCB_GetBreakpoint
{
	let mut result: seL4_TCB_GetBreakpoint = ::core::mem::zeroed();
	let tag = seL4_MessageInfo::new(InvocationLabel::TCBGetBreakpoint as seL4_Word, 0, 0, 1);
...
```
Rust generated with output arguments that are `not pass_by_reference` added.
```
#[inline(always)]
pub unsafe fn seL4_TCB_GetBreakpoint(_service: seL4_TCB, bp_num: u16, vaddr: *mut seL4_Word, type_: *mut seL4_Word, size: *mut seL4_Word, rw: *mut seL4_Word, is_enabled: *mut u8) -> seL4_TCB_GetBreakpoint
{
    let mut result: seL4_TCB_GetBreakpoint = ::core::mem::zeroed();
    let tag = seL4_MessageInfo::new(InvocationLabel::TCBGetBreakpoint as seL4_Word, 0, 0, 1t);
```

Output arguments that are not `pass_by_reference` also appear in 4 other functions, and skipping them results in code that corresponds to the C code.
These lists of parameters are in fact obtained using `syscall_stub_gen.parse_xml_file`.

Reply:
Thanks very much to looking into this. Having a comment here would be helpful that dropping certain params is really intended (and no else-branch got lost somewhere).

Reply:
Sure, right now there's a comment informing that we're dropping these params intentionally.